### DISCUSSION TITLE
[Orange Pi 5 Plus] I2S output with 40 pin GPIO header on 24.10 builds (mainline-6.11 oracular)

### DISCUSSION THREAD

**Discussion by nobara256** [Joshua-Riek/ubuntu-rockchip#1116] [Show and tell]
RK3588 has total 4 i2s interfaces (i2s0/i2s1 with 8 channels and i2s2/i2s3 with 2 channels).

Orange Pi 5 Plus GPIO Pinout diagrams on the internet don't have any specific clue if any of these i2s interfaces are exposed via the 40 pin GPIO header.

However the Orange Pi 5 Plus schematics on the official downloads has something interesting
<img src="https://github.com/user-attachments/assets/88d7746d-68b7-4995-bf8f-9098cc441e31" width=50% heigh=50%>

The Pins 40, 38, 35, 31, 12 seems to have some PCM functionalities.

Further diving into the dts files in the oracular branch

from `rk3588-base-pinctrl.dtsi` we conform that these are indeed i2s3 pins.
<img src="https://github.com/user-attachments/assets/196e46bb-3aca-4ce2-95e9-07a09e9e9eb6" width=70% height=70%>

again in `rk3588-base.dtsi` the i2s3_2c interface is already there with proper pins assigned.
<img src="https://github.com/user-attachments/assets/250d1e29-b09c-417d-88c1-0a848d047182" width=70% height=70%>

So with my zero knowledge with dts, after a lot of trial and errors, I made this overlay to add a pcm5102 node, for simple i2s output.

```
/dts-v1/;
/plugin/;

/ {
	compatible = "xunlong,orangepi-5-plus", "rockchip,rk3588";

	/* Simple I2S output can be used with with PCM5102a DAC */
	fragment@0 {
		target-path = "/";

		__overlay__ {
			pcm5102 {
				compatible = "simple-audio-card";

				status = "okay";

				simple-audio-card,name = "PCM5102";
				simple-audio-card,format = "i2s";
				simple-audio-card,bitclock-master = <&pcm5102_cpu>;
				simple-audio-card,frame-master = <&pcm5102_cpu>;

				simple-audio-card,widgets =
							"Headphone", "Headphone",
							"Speaker", "Speaker",
							"Microphone", "Microphone";

				pcm5102_cpu: simple-audio-card,cpu {
					sound-dai = <&i2s3_2ch>;
					/* system-clock-frequency = <12288000>; */
				};

				pcm5102_codec: simple-audio-card,codec {
					sound-dai = <&pcm5102_master>;
					/* system-clock-frequency = <12288000>; */
				};
			};

			/* Simple Codec PCM5102a doesn't have a I2C control interface */
			pcm5102_master: pcm5102a {
				#sound-dai-cells = <0>;
				compatible = "ti,pcm5102a";
			};
		};
	};
};
```

compile it with `device-tree-compiler`
```
dtc -@ -I dts -O dtb i2s3-overlay.dts -o i2s3-overlay.dtbo
```

copy the `.dtbo` to the path `/lib/firmware/$(uname -r)/device-tree/rockchip/overlay` (create the path if not exists) and then edit the `/etc/default/u-boot` like this.

```
## /etc/default/u-boot - configuration file for u-boot-update(8)

#U_BOOT_UPDATE="true"

#U_BOOT_ALTERNATIVES="default recovery"
#U_BOOT_DEFAULT="l0"
#U_BOOT_PROMPT="1"
#U_BOOT_ENTRIES="all"
#U_BOOT_MENU_LABEL="Debian GNU/Linux"
#U_BOOT_PARAMETERS="ro earlycon"
#U_BOOT_ROOT=""
#U_BOOT_TIMEOUT="50"
#U_BOOT_FDT=""
#U_BOOT_FDT_DIR="/lib/firmware/"
U_BOOT_FDT_OVERLAYS="device-tree/rockchip/overlay/i2s3-overlay.dtbo"
#U_BOOT_FDT_OVERLAYS_DIR="/lib/firmware/"
#U_BOOT_SYNC_DTBS="false"
```

> **Tip**
> don't forget to run `sudo u-boot-update` after saving the u-boot.

Result appear as `alsa_output.platform-pcm5102.stereo-fallback` and `Built-in Audio Stereo` alongside the onboard `es8328` that doesn't seem to work for on the mainline builds.

<img src="https://github.com/user-attachments/assets/1747ca00-bbfd-4546-9328-461469a2c0fa" width=75% height=75%>

<img src="https://github.com/user-attachments/assets/3ef8da57-4e7d-4973-9e9e-be6acd7c86b8" width=75% height=75%>

<img src="https://github.com/user-attachments/assets/b8c30cde-35bd-4b46-8b7c-97afac20a640">

With the same procedure I am able to add I2C, SPI, PWM overlays as well, which are missing from the mainline but still has every thing defined in the `rk3588-base.dtsi` & `rk3588-base-pinctrl.dtsi`.

I have been looking for i2s on Orange Pi 5 Plus for more than a year. There is very little documentation on this topic, I hope this would help someone out.

**Comment by ikey-ru** [comment]
I'm stuck here:

> copy the `.dtbo` to the path `/lib/firmware/$(uname -r)/device-tree/rockchip/overlay` (create the path if not exists) - **is ok**
 and then edit the `/etc/default/u-boot` like this. - **problems**

- My OS boots from a microSD card. I am missing a file /etc/default/u-boot and this command (u-boot-update) does not work.
I copy the `.dtbo` to the path /boot/dtb/rockchip/overlay/ and running the command `sudo orangepi-config` and activated i2s3-overlay 
Now my file /boot/orangepiEnv.txt 
```
verbosity=1
bootlogo=true
extraargs=cma=128M
overlay_prefix=rk3588
fdtfile=rockchip/rk3588s-orangepi-5-pro.dtb
rootdev=UUID=99a4ebdd-350c-4623-aa2c-d152a279947a
rootfstype=ext4
overlays=i2s3-overlay
```
I reboot the system, but the device does not appear.
pacmd list-sinks - 2 sink(s) available. <alsa_output.platform-dp0-sound.stereo-fallback> and <alsa_output.platform-es8388-sound.stereo-fallback>

I haven't connected the pcm5102 board yet. But I'm not sure that the reason is its absence. Tell me, what am I doing wrong?

**Reply by nobara256** [reply]
My procedures and results were all based on Images exclusively built by Joshua. If you are using official images from OrangePi these tricks probably won't work. Also I haven't tested them on the rockchip kernel 6.1. In theory, you should be able to use the custom overlay just fine, there was probably some thing left to do after modifying the `orangepiEnv.txt`.

**Reply by ikey-ru** [reply]
Ok, I'll try Joshua, tahnks

**Reply by nobara256** [reply]
Ok I tried to build the upstream mainline kernel and it also failed to load my .dbto as well. However if I alter the base .dtb in `/etc/default/u-boot` with the ones from Joshua's images, it works fine.

Hooking up the debug UART with 1500000 baud rate I can see it failed to load it.

```
Retrieving file: /lib/firmware/6.13.0-rc2-rockchip/device-tree/rockchip/rk3588-orangepi-5-plus.dtb
87823 bytes read in 606 ms (140.6 KiB/s)
Retrieving file: /lib/firmware/6.13.0-rc2-rockchip/device-tree/rockchip/overlay/nobara.dtbo
2449 bytes read in 519 ms (3.9 KiB/s)
failed on fdt_overlay_apply(): FDT_ERR_NOTFOUND
base fdt does did not have a /__symbols__ node
make sure you've compiled with -@
Failed to apply overlay /lib/firmware/6.13.0-rc2-rockchip/device-tree/rockchip/overlay/nobara.dtbo, skipping
Fdt Ramdisk skip relocation
No misc partition
ERROR: Did not find a cmdline Flattened Device Tree
Could not find a valid device tree
```

Look how it complains about `make sure you've compiled with -@`, while checking the Makefile, it didn't had that '-@' flag in the upstream kernel.

```
quiet_cmd_dtc = DTC $(quiet_dtb_check_tag) $@
      cmd_dtc = \
	$(HOSTCC) -E $(dtc_cpp_flags) -x assembler-with-cpp -o $(dtc-tmp) $< ; \
	$(DTC) -o $@ -b 0 $(addprefix -i,$(dir $<) $(DTC_INCLUDE)) \ // <- look here
	       $(DTC_FLAGS) -d $(depfile).dtc.tmp $(dtc-tmp) ; \
	cat $(depfile).pre.tmp $(depfile).dtc.tmp > $(depfile) \
	$(cmd_dtb_check)
```

Adding `$(DTC) -@` and then compiling the kernel (or at least dtbs) solves the issue and then it can load the overlays.


It makes sense that different images and kernels are built with different flags and thus lack absolute level of compatibility. The best way to debug would be to use the UART output to monitor the boot logs. Make sure to use a USB-TTL converter that supports upto 3Mbaud or higher like the ones with the ftdi chips our just a simple Raspberry Pi Pico.

**Comment by ikey-ru** [comment]
I specifically purchased orange5 plus but the situation has not changed. Tell me what I am doing wrong?

1. Loading and write on MicroSD [Ubuntu 22.04 LTS Desktop with Linux 5.10](https://joshua-riek.github.io/ubuntu-rockchip-download/boards/orangepi-5-plus.html)
2. nano i2s3-overlay.dts (and copy/paste your text overlay)
3. dtc -@ -I dts -O dtb i2s3-overlay.dts -o i2s3-overlay.dtbo
4. chmod u+x i2s3-overlay.dtbo
5. sudo cp i2s3-overlay.dtbo /lib/firmware/$(uname -r)/device-tree/rockchip/overlay/i2s3-overlay.dtbo
6. sudo nano /etc/default/u-boot (and paste U_BOOT_FDT_OVERLAYS="device-tree/rockchip/overlay/i2s3-overlay.dtbo")
7. sudo u-boot-update
8. sudo reboot
`wpctl status` and `pactl list sinks` give nothing to indicate the presence of pcm5102
I connected pin PCM5102 BCK to 12 pin Orange PI, LCK to 38 pin OP, Din to 40 pin.

**Reply by nobara256** [reply]
Try running `dmesg | grep -i pcm` and see if anything shows up. If the driver has no conflicts and issues, then nothing should show up. If you have a 1.5Mbaud UART-USB converter then try hooking up the debug pins (3 pins near power port) and monitor the initial boot logs before kernel loads up.

Also checkout the 24.10 image with linux-6.11 and verify the setup.

**Reply by ikey-ru** [reply]
```
ik@ik-desktop:~$ dmesg | grep -i pcm
[   17.775706] rockchip-i2s-tdm fe470000.i2s: Register PCM for TRCM mode
```
I don't really understand what this means.(

24.10 image with linux-6.11 - I checked. The result is the same.
1.5Mbaud UART-USB converter - I'll try to find it.

**Reply by nobara256** [reply]
Try the overlay below with 24.10 "server" images.

```
/dts-v1/;
/plugin/;

/ {
	compatible = "xunlong,orangepi-5-plus", "rockchip,rk3588";

	/* I2C2_M0 */
	fragment@0 {
		target = <&i2c2>;

		__overlay__ {
			status = "okay";
			pinctrl-names = "default";
			pinctrl-0 = <&i2c2m0_xfer>;

			#address-cells = <1>;
        	#size-cells = <0>;

			ina219@40 {
				compatible = "ti,ina219";
				label = "vdd_12v0";
				reg = <0x40>;
				shunt-resistor = <100000>;
			};
		};
	};

	/* SPI0_M2 */
	fragment@1 {
		target = <&spi0>;

		__overlay__ {
			status = "okay";
			#address-cells = <1>;
			#size-cells = <0>;
			pinctrl-names = "default";
			pinctrl-0 = <&spi0m2_cs0 &spi0m2_cs1 &spi0m2_pins>;
			max-freq = <50000000>;

			spidev@0 {
				compatible = "rohm,bh2228fv";
				status = "okay";
				reg = <0>;
				spi-max-frequency = <50000000>;
			};
		};
	};

	/* PWM14_M2 */
	fragment@2 {
		target = <&pwm14>;

		__overlay__ {
			status = "okay";
			pinctrl-0 = <&pwm14m2_pins>;
		};
	};

	/* I2S3 2 channel */
	fragment@3 {
		target = <&i2s3_2ch>;

		__overlay__ {
			pinctrl-names = "default";
			pinctrl-0 = <&i2s3_lrck
						 &i2s3_sclk
						 &i2s3_sdi
						 &i2s3_sdo>;
			/* assigned-clock-rates = <49152000>, <24576000>, <12288000>; */
			status = "okay";
		};
	};

	/* Simple I2S output can be used with with PCM5102a DAC */
	fragment@4 {
		target-path = "/";

		__overlay__ {
			pcm5102 {
				compatible = "simple-audio-card";

				status = "okay";

				simple-audio-card,name = "PCM5102";
				simple-audio-card,format = "i2s";
				simple-audio-card,mclk-fs = <256>;
				simple-audio-card,bitclock-master = <&pcm5102_cpu>;
				simple-audio-card,frame-master = <&pcm5102_cpu>;

				simple-audio-card,widgets =
							"Headphone", "Headphone Jack",
							"Microphone", "Microphone Jack";

				pcm5102_cpu: simple-audio-card,cpu {
					sound-dai = <&i2s3_2ch>;
					/* system-clock-frequency = <12288000>; */
				};

				pcm5102_codec: simple-audio-card,codec {
					sound-dai = <&pcm5102_master>;
					/* system-clock-frequency = <12288000>; */
				};
			};

			/* Simple Codec PCM5102a doesn't have a I2C control interface */
			pcm5102_master: pcm5102a {
				#sound-dai-cells = <0>;
				compatible = "ti,pcm5102a";
			};
		};
	};
};
```

It has a lot of other stuff but this is "exactly" what I am using right now. One more thing (shouldn't be the issue) all of my tests are done on the server images by far, so try one orangepi-5-plus ubuntu 24.10 server image.

**Reply by nobara256** [reply]
Now that I look close my original post lacks this fragment below!! (Sorry, I'm an Idiot.)

```
	/* I2S3 2 channel */
	fragment@3 {
		target = <&i2s3_2ch>;

		__overlay__ {
			pinctrl-names = "default";
			pinctrl-0 = <&i2s3_lrck
						 &i2s3_sclk
						 &i2s3_sdi
						 &i2s3_sdo>;
			/* assigned-clock-rates = <49152000>, <24576000>, <12288000>; */
			status = "okay";
		};
	};
```

Without this, the underlying i2s channel won't work and that might be the issue here. Can you put the full overlay file and give it a try on the 24.10 images.

**Reply by ikey-ru** [reply]
It's gotten much better.
The device has appeared. `dmesg | grep -i pcm` gives no errors.
Although, I can't understand why there is no sound at the output, but maybe it's a connection problem.
I'll look into it. Thanks for your help. There really is very little information on this topic.

The status now shows this:

```
ubuntu@ubuntu:~$ wpctl status
PipeWire 'pipewire-0' [1.2.4, ubuntu@ubuntu, cookie:115226102]
 └─ Clients:
        33. WirePlumber                         [1.2.4, ubuntu@ubuntu, pid:1128]
        37. pipewire                            [1.2.4, ubuntu@ubuntu, pid:1129]
        47. WirePlumber [export]                [1.2.4, ubuntu@ubuntu, pid:1128]
        53. ffplay                              [1.2.4, ubuntu@ubuntu, pid:4007]
        72. wpctl                               [1.2.4, ubuntu@ubuntu, pid:4201]

Audio
 ├─ Devices:
 │      62. Built-in Audio                      [alsa]
 │      63. Built-in Audio                      [alsa]
 │  
 ├─ Sinks:
 │  *   64. Built-in Audio Stereo               [vol: 0.40]
 │      66. Built-in Audio Stereo               [vol: 0.40]
 │  
 ├─ Sources:
 │  *   65. Built-in Audio Stereo               [vol: 1.00]
 │  
 ├─ Filters:
 │  
 └─ Streams:
        54. ffplay                                                      
             55. output_FL       > fe4a0000.i2s-pcm5102a-hifi pcm5102a-hifi-0:playback_FL	[active]
             56. output_FR       > fe4a0000.i2s-pcm5102a-hifi pcm5102a-hifi-0:playback_FR	[active]

Video
 ├─ Devices:
 │      48. rockchip,rk3568-vpu-dec             [v4l2]
 │      49. rockchip-rga                        [v4l2]
 │      50. rockchip,rk3588-vepu121-enc         [v4l2]
 │      51. rkvdec2                             [v4l2]
 │      52. rockchip,rk3588-av1-vpu-dec         [v4l2]
 │  
 ├─ Sinks:
 │  
 ├─ Sources:
 │  
 ├─ Filters:
 │  
 └─ Streams:

Settings
 └─ Default Configured Devices:
```
```
ubuntu@ubuntu:~$ pactl list sinks
Sink #69
	State: RUNNING
	Name: alsa_output.platform-pcm5102.stereo-fallback
	Description: Built-in Audio Stereo
	Driver: PipeWire
	Sample Specification: s32le 2ch 48000Hz
	Channel Map: front-left,front-right
	Owner Module: 4294967295
	Mute: no
	Volume: front-left: 26214 /  40% / -23.88 dB,   front-right: 26214 /  40% / -23.88 dB
	        balance 0.00
	Base Volume: 65536 / 100% / 0.00 dB
	Monitor Source: alsa_output.platform-pcm5102.stereo-fallback.monitor
	Latency: 0 usec, configured 0 usec
	Flags: HARDWARE DECIBEL_VOLUME LATENCY 
	Properties:
		alsa.card = "0"
		alsa.card_name = "PCM5102"
		alsa.class = "generic"
		alsa.device = "0"
		alsa.driver_name = "snd_soc_simple_card"
		alsa.id = "PCM5102"
		alsa.long_card_name = "PCM5102"
		alsa.name = "fe4a0000.i2s-pcm5102a-hifi pcm5102a-hifi-0"
		alsa.resolution_bits = "16"
		alsa.subclass = "generic-mix"
		alsa.subdevice = "0"
		alsa.subdevice_name = "subdevice #0"
		alsa.sync.id = "00000000:00000000:00000000:00000000"
		api.alsa.card.longname = "PCM5102"
		api.alsa.card.name = "PCM5102"
		api.alsa.path = "hw:0"
		api.alsa.pcm.card = "0"
		api.alsa.pcm.stream = "playback"
		audio.channels = "2"
		audio.position = "FL,FR"
		card.profile.device = "1"
		device.api = "alsa"
		device.class = "sound"
		device.id = "62"
		device.profile.description = "Stereo"
		device.profile.name = "stereo-fallback"
		device.routes = "1"
		factory.name = "api.alsa.pcm.sink"
		media.class = "Audio/Sink"
		device.description = "Built-in Audio"
		node.name = "alsa_output.platform-pcm5102.stereo-fallback"
		node.nick = "fe4a0000.i2s-pcm5102a-hifi pcm5102a-hifi-0"
		node.pause-on-idle = "false"
		object.path = "alsa:acp:PCM5102:1:playback"
		port.group = "playback"
		priority.driver = "1000"
		priority.session = "1000"
		factory.id = "19"
		clock.quantum-limit = "8192"
		client.id = "47"
		node.driver = "true"
		node.loop.name = "data-loop.0"
		library.name = "audioconvert/libspa-audioconvert"
		object.id = "64"
		object.serial = "69"
		node.max-latency = "16384/48000"
		api.alsa.period-size = "1024"
		api.alsa.period-num = "32"
		api.alsa.headroom = "0"
		api.acp.auto-port = "false"
		api.alsa.card = "0"
		api.alsa.use-acp = "true"
		api.dbus.ReserveDevice1 = "Audio0"
		api.dbus.ReserveDevice1.Priority = "-20"
		device.bus_path = "platform-pcm5102"
		device.enum.api = "udev"
		device.form_factor = "internal"
		device.icon_name = "audio-card-analog"
		device.name = "alsa_card.platform-pcm5102"
		device.nick = "PCM5102"
		device.plugged.usec = "16734757"
		device.subsystem = "sound"
		sysfs.path = "/devices/platform/pcm5102/sound/card0"
		device.string = "0"
	Ports:
		analog-output: Analog Output (type: Analog, priority: 9900, availability unknown)
	Active Port: analog-output
	Formats:
		pcm
		```

**Reply by nobara256** [reply]
Great, the sink is recognized and running, you can get the PINOUT from the schematics picture. Give us an update whenever you get the some sound output.

There is really very little information on this topic online. And because every board manufacturer has different device tree files, it's kind of difficult to work with. Hope this actually help others looking into this.

**Reply by cselip** [reply]
thank you very much for this thread, it has been super helpful

any tips on how to get this working with armbian on 6.1.115-vendor-rk35xx, with orange pi 5 pro ? I confirmed it matches your first schematic

I tried your most recent version, except with modifying the armbianEnv.txt and adding user_overlays= since I am not on the ubuntu image,

I know it is applying because I get 
```
[    7.695551] rockchip-pinctrl pinctrl: pin gpio3-2 already requested by fe2d0000.mmc; cannot claim for fe4a0000.i2s
[    7.695560] rockchip-pinctrl pinctrl: pin-98 (fe4a0000.i2s) status -22
[    7.695568] rockchip-pinctrl pinctrl: could not request pin 98 (gpio3-2) from group i2s3-lrck  on device rockchip-pinctrl
[    7.695576] rockchip-i2s fe4a0000.i2s: Error applying setting, reverse things back
```
so I had to add a fragment to disable that node, but after fixing that I don't really see anything interesting when I look through dmesg and grep for pcm like you guys did

any idea what I might have to change? thanks

**Comment by maxycatz** [comment]
Coming to this late, but wondering if the above is effective to get I2S output via the GPIO.  Is it confirmed?

**Reply by maxycatz** [reply]
Yes, it does; just ran it.  Thanks for solving this, nobara256.
